armory/utils/export.py

```diff
@@ -92,369 +77,579 @@ def _make_output_dir(self):
-                    s=f"{x_i.shape[2]}x{x_i.shape[1]}",
-                )
-                .output(
-                    os.path.join(self.output_dir, folder, "video_adversarial.mp4"),
-                    pix_fmt="yuv420p",
-                    vcodec="libx264",
-                    r=self.context.frame_rate,
-                )
-                .overwrite_output()
-                .run_async(pipe_stdin=True, quiet=True)
+        :param x_i: floating point np array of shape (num_frames, H, W, C=3) in [0.0, 1.0]
+        :param with_boxes: boolean indicating whether to display bounding boxes
+        :param y_i: ground-truth label dict
+        :param y_i_pred: predicted label dict
+        :return: List[PIL.Image.Image] of length equal to num_frames
+        """
+        if not with_boxes:
+            return super().get_sample(x_i)
+
+        if y_i is None and y_i_pred is None:
+            raise TypeError("Both y_i and y_pred are None, but with_boxes is True.")
+        if x_i.min() < 0.0 or x_i.max() > 1.0:
+            log.warning("video out of expected range. Clipping to [0,1]")
+
+        pil_frames = []
+        for n_frame, x_frame, in enumerate(x_i):
+            pixels = np.uint8(np.clip(x_frame, 0.0, 1.0) * 255.0)
+            image = Image.fromarray(pixels, "RGB")
+            box_layer = ImageDraw.Draw(image)
+            if y_i is not None:
+                bbox_true = y_i["boxes"][n_frame].astype("float32")
+                box_layer.rectangle(bbox_true, outline="red", width=2)
+            if y_i_pred is not None:
+                bbox_pred = y_i_pred["boxes"][n_frame]
+                box_layer.rectangle(bbox_pred, outline="white", width=2)
+            pil_frames.append(image)
+
+        return pil_frames
+
+
+class AudioExporter(SampleExporter):
+    def __init__(self, base_output_dir, sample_rate):
+        self.sample_rate = sample_rate
+        super().__init__(base_output_dir)
+
+    def _export(
+        self, x, x_adv=None, y=None, y_pred_adv=None, y_pred_clean=None, **kwargs
+    ):
+        for i, x_i in enumerate(x):
+            self._export_audio(x_i, type="benign")
+
+            if x_adv is not None:
+                x_i_adv = x_adv[i]
+                self._export_audio(x_i_adv, type="adversarial")
+
+            self.saved_samples += 1
+        self.saved_batches += 1
+
+    def _export_audio(self, x_i, type="benign"):
+        if type not in ["benign", "adversarial"]:
+            raise ValueError(
+                f"type must be one of ['benign', 'adversarial'], received '{type}'."
             )
 
-            for n_frame, (x_frame, x_adv_frame) in enumerate(zip(x_i, x_adv_i)):
+        if x_i.min() < -1.0 or x_i.max() > 1.0:
+            log.warning("input out of expected range. Clipping to [-1, 1]")
+
+        wavfile.write(
+            os.path.join(self.output_dir, f"{self.saved_samples}_{type}.wav"),
+            rate=self.sample_rate,
+            data=np.clip(x_i, -1.0, 1.0),
+        )
```

Comment:
Audio inputs are limited to the [-1.0, 1.0] range, but this is not true after adversarial additions. Instead of clipping in this case, I would normalize by the max value, `np.abs(x_i).max()` and have the warning say that.

We may want to consider quantizing the output to np.int16 after dividing by 2**15 (see datasets.AudioContext).

We should also add a check for infinite and NaN values and throw a warning if they are there. (Possibly using `nan_to_num`).

Reply:
I assume you mean ` if not np.isfinite(x_i_copy).all():`, but this makes sense. Just pushed an update

Reply:
R.e. `We may want to consider quantizing the output to np.int16 after dividing by 2**15 (see datasets.AudioContext)`:

Are you saying what gets passed into `wavfile.write()` should be scaled back to int's? From looking at the datasets code, this is my understanding:

1. raw audio data is in `int` form
2. during preprocessing we divide by 2 ** 15 converting it to `float`

The `x_i` that `_export_audio()` sees is this scaled down `float` version, and currently it passes this along to `wavfile.write()`.

If you call `get_sample()`, on the other hand, per your recommendation earlier I'm scaling the float array back to ints after multiplying by 2 ** 15. Does that all make sense? Do you have thoughts on whether the Audio `get_sample()` method should

1. take in a `dataset_context` arg, from which `dataset.quantization` is obtained
2. take in a `quantization` arg where you pass the int used to scale back up the sample
3. take no extra arg, and always scale up with 2 ** 15

Reply:
I would probably pass in the dataset_context. Then you can at least assert that the context input_type is np.int16 and the quantization is 2**15.
https://github.com/twosixlabs/armory/blob/develop/armory/data/datasets.py#L720

One corner case of note: after scaling up, you'll need to clip any 2**15 values to 2**15 - 1. (You can just clip by context.input_min and context.input_max).